### ISSUE TITLE
Reading Single Input or Output

### ISSUE THREAD

**Issue by ufukizgi** [gijzelaerr/python-snap7#53]
Hello,

I am really not very good at PLC also Python :smile: 
I am working on python-snap7 on a few days. I am able to connect, read data and convert to boolean but I can not understand which one is the boolean that I am looking for.
In brief, how can I read a specific input or output from PLC? (e.g. : I3.2 or Q0.4 etc.)
I will be glad if you explain or share some links. Because I couldn't find any.

Thanks,
Ufuk.

**Comment by ufukizgi** [comment]
That is what I mean. I can't find my fault. Bits matchs on Snap7 Client and on Step7 but they are not same as the booleans my codes returns. 

![snap7](https://cloud.githubusercontent.com/assets/14932066/10256154/efad89b8-6957-11e5-9652-3c1e6b47cbb3.PNG)

**Comment by spreeker** [comment]
ab_read(0,8) is reading 8 bytes and puts it in an bytearray.

    bytearray = ab_read(0, 1)    # read the fist bite 

    for index in range(8):           # loop over for each bit
         byte_with_bools = 0
         print getbool(bytearray, byte_with_bools, index)    #

**Comment by ufukizgi** [comment]
Thank you but unfortunately nothing changed. Your advice as same as mine indeed. Only the size of bytearray changed. 
I have read the source. I am in doubt about bit order on bytearray and PLC are not same. If you have a look at ss, values are changing but they should be same on VAT_3 screen.

![ss](https://cloud.githubusercontent.com/assets/14932066/10263282/fa1fb3f0-69f0-11e5-81f8-a57f3ddb40e8.JPG)

**Comment by spreeker** [comment]
Yes the byte order could be different... Maybe the magic variable indicating the correct position in the plc of the ab memory block could be wrong.  You could check the variable used in the python source compared to the one used by the snap7 or use the read area function. If you know the exact memory location.

**Comment by ufukizgi** [comment]
Eureka!!

Finally I found my mistake. I was using ab_read() and I was looking for match in Inputs  :sob: 
In order to find I changed the codes little bit but at least I have found match in Outputs.
But if these are outputs, where are inputs? How can I read them?

Thank you for your help, spreeker.

![1510040220](https://cloud.githubusercontent.com/assets/14932066/10265585/d3d3d106-6a3e-11e5-96ee-f146a1eac209.PNG)

**Comment by spreeker** [comment]
Read the source of the client clode. The read_area function can do it all. 

      def read_area(self, area, dbnumber, start, size):
        """This is the main function to read data from a PLC.
        With it you can read DB, Inputs, Outputs, Merkers, Timers and Counters.

        :param dbnumber: The DB number, only used when area= S7AreaDB
        :param start: offset to start writing
        :param size: number of units to read
        """
        assert area in snap7.snap7types.areas.values()
        wordlen = snap7.snap7types.S7WLByte
        type_ = snap7.snap7types.wordlen_to_ctypes[wordlen]
        logging.debug("reading area: %s dbnumber: %s start: %s: amount %s: "
                      "wordlen: %s" % (area, dbnumber, start, size, wordlen))
        data = (type_ * size)()
        result = self.library.Cli_ReadArea(self.pointer, area, dbnumber, start,
                                           size, wordlen, byref(data))
        check_error(result, context="client")
        return bytearray(data)

**Comment by ufukizgi** [comment]
Ah OK thanks.

**Comment by gijzelaerr** [comment]
@ufukizgi did this solve your problem?

**Comment by ufukizgi** [comment]
Yes, I am abble to read inputs also. I attached the picture below in order to be an example for newbies as me. :smile: 

Thanks for help.

![ssinput](https://cloud.githubusercontent.com/assets/14932066/10280569/aff27416-6b76-11e5-9fb8-16f480b043d1.JPG)

**Comment by lupodellasleppa** [comment]
Hello

first of all thanks for all the great work, been using this library on my work projects and works like a charm.

I'm reopening this issue tho, just for a couple of questions.

When reading outputs, we have `ab_read` method.
But when trying to read using `read_area`, I couldn't find the correct `snap7types` area code to use.

Is there anywhere we can find a thorough mapping of all PLC areas, and their snap7 relatives? I've looked everywhere throughout the documentation and source code without any success.

for now, what i came up with is the following:

```
S7AreaPE = Inputs (I)
S7AreaPA = Outpus (Q)
S7AreaMK = Merkers (M)
S7AreaDB = Data Blocks (DB)
S7AreaCT = Counters (C)
S7AreaTM = Timers (T)
```

Is this correct? I've reported in parenthesis the letter you find in the Step7 program.

Some more questions:
Why use different names like AB and PA?
Is there a way to read blocks such as OB, FB, SFC etc.? (not just their size)

Sorry for my silly questions, but my knowledge of PLCs is very limited.

Thank you very much for all your work
